## Dataset of Used Cars from the German eBay Website

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Brief First Observations (only from the data printed above):

Null objects: 

- 6   vehicleType          44905 non-null  object
- 8   gearbox              47320 non-null  object
- 10  model                47242 non-null  object
- 13  fuelType             45518 non-null  object
- 15  notRepairedDamage    40171 non-null  object

Price range: 1k-22.9k

Dates: All 2016 and in March and April.

Dtypes: int and object types. No float types.

Columns: The dateCrawled and name columns are rather messy and look like they'll need work.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
new_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen']

autos.columns = new_columns

In [7]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

### The above three cells, explained:

I changed the column names from camelcase to snakecase. I am not entirely sure why, but I believe it'll make coding easier? It's possible the Python language prefers snakecase when coding and so it'll make things run smoother. I think it made everything look neat.

In [8]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-04 16:40:33  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [9]:
autos.seller.value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [10]:
autos.offer_type.value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [11]:
autos.abtest.value_counts()

test       25756
control    24244
Name: abtest, dtype: int64

In [12]:
autos.nr_pictures.value_counts()

0    50000
Name: nr_pictures, dtype: int64

## All the above four columns are candidates to be dropped

They mostly have one value and may not be useful to analysis

In [13]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  object
 5   abtest              50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Columns with Numeric Data Stored As Text

price, odometer, postal code

And potentially the data_crawled, ad_created, and last_seen columns? 

We will remove the nr_pictures column

In [14]:
autos.price = autos.price.str.replace("$", "")
autos.price = autos.price.str.replace(",", "")
autos.price = autos.price.astype(int)

In [15]:
autos.price.dtype

dtype('int64')

In [16]:
autos.odometer

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [17]:
autos.odometer = autos.odometer.str.replace("km", "")
autos.odometer = autos.odometer.str.replace(",", "")
autos.odometer = autos.odometer.astype(int)

In [18]:
autos = autos.rename({"odometer": "odometer_km"}, axis = 1)

In [19]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [20]:
autos.price.unique().shape

(2357,)

In [21]:
autos.price.describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [22]:
autos.price.value_counts().sort_index(ascending = True).tail(15)

265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64

In [23]:
autos.odometer_km.value_counts().sort_index(ascending = True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [24]:
autos = autos[autos["price"].between(0, 350000)]

The instructions called for removing outliers. I decided not too. I'm interested in working with a dataset that is completely representative of what was actually on the website at the time. Whether that be a car that was listed for 0, or a car that was listed for 99,999,999.

In [25]:
autos.date_crawled.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05 14:06:30    0.00002
2016-03-05 14:06:40    0.00002
2016-03-05 14:07:04    0.00002
2016-03-05 14:07:08    0.00002
2016-03-05 14:07:21    0.00002
                        ...   
2016-04-07 14:30:09    0.00002
2016-04-07 14:30:26    0.00002
2016-04-07 14:36:44    0.00002
2016-04-07 14:36:55    0.00002
2016-04-07 14:36:56    0.00002
Name: date_crawled, Length: 48200, dtype: float64

### Observations of the above 

- Date range 03-05 '16 to 04-07 '16
- No value counts over .00006. Minimum is .00002
- 48213 values

In [26]:
autos.ad_created.value_counts(normalize = True, dropna = False).sort_index()

2015-06-11 00:00:00    0.000020
2015-08-10 00:00:00    0.000020
2015-09-09 00:00:00    0.000020
2015-11-10 00:00:00    0.000020
2015-12-05 00:00:00    0.000020
                         ...   
2016-04-03 00:00:00    0.038931
2016-04-04 00:00:00    0.036850
2016-04-05 00:00:00    0.011843
2016-04-06 00:00:00    0.003261
2016-04-07 00:00:00    0.001280
Name: ad_created, Length: 76, dtype: float64

### Observations of the above
- 76 Values
- Majority of ads created between 3-12 '16 and 04-03 '16
- Max value .03892, Min value .00002. Difference = .03890
- Date Range 06-11 '15 to 04-07 '16

In [27]:
autos.last_seen.value_counts(normalize = True, dropna = False).sort_index()

2016-03-05 14:45:46    0.00002
2016-03-05 14:46:02    0.00002
2016-03-05 14:49:34    0.00002
2016-03-05 15:16:11    0.00002
2016-03-05 15:16:47    0.00002
                        ...   
2016-04-07 14:58:44    0.00006
2016-04-07 14:58:45    0.00002
2016-04-07 14:58:46    0.00002
2016-04-07 14:58:48    0.00006
2016-04-07 14:58:50    0.00008
Name: last_seen, Length: 39472, dtype: float64

### Observations of the above

- Date Range: 03-05 '16 to 04'07 '16
- 39481 Values
- .0016 is the max value count, .00002 is the minimum
     - It makes sense that the most "last_seen" ads would be on one of the last days recorded. 

In [28]:
autos.registration_year.describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### Observations on the above

There seems to be some anomalies and errors in the data. For example, a minimum year of 1000 and a maximum year of 9999 do not accurately represent when possible registration years are for this dataset. Unless...one car might be from the future. The average registration year is 2005, with the 25% percentile being 1999 and the 75% percentile being 2008.

The lowest and highest acceptable values for the registration_year columns are: 

1. When the first car was registered, which may be 1901 from a quick google search. 
2. The last date when a listing was seen from the dataset which was 04/07/2016. How could it be possible for a car to be registered past that date? The date didn't happen yet, so a car could not be registered past that date and be listed. 

Therefore, the lowest and highest acceptable values are 1901 and 2016.

In [29]:
autos = autos[autos["registration_year"].between(1901, 2016)]

In [30]:
autos["registration_year"].describe()

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [31]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [32]:
autos.date_crawled.value_counts(normalize = True, dropna = False)

2016-04-04 16:40:33    0.000062
2016-03-29 23:42:13    0.000062
2016-03-05 16:57:05    0.000062
2016-03-19 17:36:18    0.000062
2016-03-08 10:40:35    0.000062
                         ...   
2016-03-21 22:54:25    0.000021
2016-03-29 16:36:24    0.000021
2016-03-20 13:51:39    0.000021
2016-03-21 20:53:15    0.000021
2016-03-10 21:39:43    0.000021
Name: date_crawled, Length: 46363, dtype: float64

### Observations on the above

- Date range: 2016-03-05 to 2016-04-07
- Max value: .000062, Min Value: .000021
- Length 46375 
- dtype: float64

In [33]:
autos.ad_created.value_counts(normalize = True, dropna = False).head(10)

2016-04-03 00:00:00    0.039112
2016-03-20 00:00:00    0.038008
2016-03-21 00:00:00    0.037675
2016-04-04 00:00:00    0.036946
2016-03-12 00:00:00    0.036529
2016-04-02 00:00:00    0.035134
2016-03-14 00:00:00    0.035030
2016-03-07 00:00:00    0.034947
2016-03-28 00:00:00    0.034738
2016-03-29 00:00:00    0.034155
Name: ad_created, dtype: float64

### Observations on the above

- Length 74, meaning there are 74 different dates represented 
- The top 4 dates where the most ads were created are 04-03, 03-20, 03-21, and 04-04. This leads me to think that there may have been something going on during those two days, such as a promotion. It turns out that 04-03 and 04-04 are a Sunday and Monday, as well as 03-20, and 03-21. Maybe the day of the week has an impact. Maybe promotions occur during those days. Maybe it's seasonal and these are typical weeks when people tend to buy used cars. Why these dates? Why these days? There could be a lot of reasons. It'd be interesting to see typical buyer behavior for used cars across multiple datasets. 

In [34]:
autos.last_seen.value_counts(normalize = True, dropna = False)

2016-04-07 06:17:27    0.000167
2016-04-07 03:16:17    0.000146
2016-04-06 21:17:51    0.000146
2016-04-06 06:17:24    0.000146
2016-04-06 10:17:12    0.000125
                         ...   
2016-03-12 23:46:06    0.000021
2016-03-24 08:16:35    0.000021
2016-03-20 04:48:17    0.000021
2016-03-15 17:44:01    0.000021
2016-04-01 17:50:19    0.000021
Name: last_seen, Length: 38093, dtype: float64

### Observations on the above

- Most ads were last seen on the last dates recorded for the datasets
- Float type datsets
- Length 38102: Thats the amount of times where ads were last seen

In [35]:
autos.registration_year.value_counts(normalize = True)

2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

### Observations on the above

The majority of veichles were registered between the dates 1999 and 2005. These registration dates makeup 5.6% - 6.9% of the dataset, depending on the date. The minority of veichles were registered between the dates 1952 and 1939. These dates all take the same value of .00021% of the dataset. 

In [36]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [37]:
brand_types = autos.brand.value_counts(normalize = True)
common_brands = brand_types[brand_types > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


### Description of Brand Data and Aggregation Decision

The data represents the distribution of the brands in the dataset. The top 3 are volkswagen, bmw, and opel.

I've chosen to aggregate the dataset by cars that take up atleast 5% of the dataset, as they offered as a recommendation. 

In [38]:
aggregated_brands = {}


for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    aggregated_brands[brand] = int(mean_price)
    
print(aggregated_brands)

{'volkswagen': 5231, 'bmw': 8102, 'opel': 2876, 'mercedes_benz': 8485, 'audi': 9093, 'ford': 3652}


### Analysis of the above results

The Audi has the highest mean price, followed by BMW, and then Ford. 


In [39]:
average_mileage = {}
for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    avg_mileage = brand_only["odometer_km"].mean()
    average_mileage[brand] = int(avg_mileage)
    
print(average_mileage)

{'volkswagen': 128724, 'bmw': 132431, 'opel': 129223, 'mercedes_benz': 130856, 'audi': 129287, 'ford': 124068}


In [40]:
avg_price = pd.Series(aggregated_brands)
print(avg_price)

volkswagen       5231
bmw              8102
opel             2876
mercedes_benz    8485
audi             9093
ford             3652
dtype: int64


In [41]:
avg_mileage = pd.Series(average_mileage)
print(avg_mileage)

volkswagen       128724
bmw              132431
opel             129223
mercedes_benz    130856
audi             129287
ford             124068
dtype: int64


In [42]:
df = pd.DataFrame(avg_price, columns = ['mean_price'])
print(df)

               mean_price
volkswagen           5231
bmw                  8102
opel                 2876
mercedes_benz        8485
audi                 9093
ford                 3652


In [43]:
df['average_mileage'] = avg_mileage

In [49]:
df

mean_price  average_mileage
volkswagen           5231           128724
bmw                  8102           132431
opel                 2876           129223
mercedes_benz        8485           130856
audi                 9093           129287
ford                 3652           124068

### Analysis of the above

The car with the highest average mileage are bmw, mercedez_benz, and Audi. These cars also have the highest prices in the dataset. It seems then that average_mileage does not negatively effect the mean_price of the car. You could technically make the argument, based on this data, that a higher average_mileage will therefore make the price of the car increase in value. However, we know that not to be the case in the real world.

I would then draw the assumption that the mean_price is more affected by the brand of the car rather than the mileage.

In [50]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50